packages/data/addon/services/bard-metadata.js

```diff
@@ -51,44 +54,56 @@ export default class BardMetadataService extends Service {
     this._serializer = owner.lookup('serializer:bard-metadata');
   }
 
+
   /**
    * @method loadMetadata
    * Fetches metadata from the bard WS using the metadata adapter and loads into keg
    *
    * @param {Object} options - options object used by the adapter
    * @returns {Promise} promise that loads metadata
    */
-  loadMetadata(options = {}) {
+  async loadMetadata(options = {}) {
     const dataSource = options.dataSourceName || getDefaultDataSourceName();
     //fetch metadata from WS if metadata not yet loaded
     if (!this.loadedDataSources.includes(dataSource)) {
-      return this._adapter
-        .fetchAll(
-          'table',
-          assign(
-            {
-              query: { format: 'fullview' }
-            },
-            options
-          )
+      const payload = await this._adapter.fetchAll(
+        'table',
+        assign(
+          {
+            query: { format: 'fullview' }
+          },
+          options
         )
-        .then(payload => {
-          //normalize payload
-          payload.source = dataSource;
-          let metadata = this._serializer.normalize(payload);
-
-          if (!(this.isDestroyed || this.isDestroying)) {
-            //create metadata model objects and load into keg
-            this._loadMetadataForType('table', metadata.tables, dataSource);
-            this._loadMetadataForType('dimension', metadata.dimensions, dataSource);
-            this._loadMetadataForType('timeDimension', metadata.timeDimensions, dataSource);
-            this._loadMetadataForType('metric', metadata.metrics, dataSource);
-
-            this.loadedDataSources.push(dataSource);
-          }
-        });
+      );
+
+      //normalize payload
+      payload.source = dataSource;
+      const metadata = this._serializer.normalize(payload);
+
+      // If metricFunctions are provided in an endpoint, fetch them, normalize them, and then load them into the keg
+      let metricFunctionMetadata = metadata.metricFunctions;
+      if (metadata.metricFunctions === null) {
+        const fetchedMetricFunctions = (await this._adapter.fetchAll('metricFunction', options)) || [];
+        metricFunctionMetadata =
+          getOwner(this)
+            .lookup(`serializer:metadata/metric-function`)
+            ?.normalize({ 'metric-functions': fetchedMetricFunctions }, dataSource) || fetchedMetricFunctions;
+      }
```

Comment:
Should this be done in the `_serializer.normalize` method?

Reply:
I don't think so because this step can only be done after a fetch has been sent to the metricFunctions endpoint. We don't know if we need to send that request until after the bard-metadata serializer's `normalize` is finished